apps/heft/src/metrics/MetricsCollector.ts

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as os from 'os';
+import { AsyncParallelHook, SyncHook } from 'tapable';
+import { performance } from 'perf_hooks';
+
+/**
+ * @public
+ */
+export interface IMetricsData {
+  command: string;
+  commandTime: number;
+  machineOs: string;
+  machineArch: string;
+  machineCores: number;
+  machineProcessor: string;
+  machineTotalMemory: number;
+}
+
+/**
+ * Tap these hooks to record events, to a file, for example.
+ *
+ * @public
+ */
+export class MetricsCollectorHooks {
+  /**
+   * This hook is called when an event is recorded.
+   */
+  public recordEvent: SyncHook<string, IMetricsData> = new SyncHook<string, IMetricsData>([
+    'eventName',
+    'metricsData'
+  ]);
+
+  /**
+   * This hook is called when events should be flushed
+   */
+  public flush: AsyncParallelHook = new AsyncParallelHook();
+
+  /**
+   * This hook is called when events should be flushed and no more events will be logged.
+   */
+  public flushAndTeardown: AsyncParallelHook = new AsyncParallelHook();
+}
+
+/**
+ * @internal
+ */
+export interface IPerformanceData {
+  duration: number;
```

Comment:
Specify units, for example `durationMs`.

Also, "duration" of what?  Should it be something like `taskTotalExecutionMs`?